jep/0000/README.adoc

```diff
@@ -0,0 +1,264 @@
+Pushing into maven repo is fine, although indexing time is big.
+Will need to mirror into Azure CDN at some point.
+
+The BoM needs to be translated to take advantage of existing infrastructure.
+
+* Essentials needs to generate a diff for upgrades and download the bits
+* Existing ATH uses a war binary that can be built from BoM
+
+Packages should be trackable back to a commit id, but during development the BoM should point to branches or tags, so we would have a BoM as input and a effective BoM as output with all the git refs resolved to specific commit ids.
+
+=== Workflow
+
+Core, libraries and plugins could have a BoM in the master branch.
+This BoM should point to master of core, master of libraries.
+There is an implicit dependency to the current branch of the current git repository.
+
+The BoM should be processed to update the Maven pom on change events.
+
+==== Changes to Core
+
+Developer needs some changes to core or Jenkins internal library or component for downstream use.
+
+* Create PR against core/library.
+** PR builder in core/library will publish the artifacts using the git commit id as part of the identifier, for downstream consumption.
+** Would also publish the result BoM with all git references converted to ids.
+* Create PR against a plugin
+** This PR would include a modified BoM pointing to the PR branch of core/library.
+* PR builder for plugin/library would process the BoM, creating as output:
+** A Jenkins package (WAR and/or Docker image) with the dependencies stated in the BoM.
+** The realized BoM, with all git references converted to ids.
+** Dependencies would not need to be rebuilt as they were already published.
+* Tests run against this output
+* Output artifact is published to repository, for downstream consumption in a way that can be fetched by git commmit id.
+
+==== Changes to Plugins
+
+Same as above but instead of pointing to core/library PR, pointing to master.
+
+==== Continuous Delivery
+
+Changes in master of core and libraries should trigger downstream rebuilds of plugins.
+Given the amount of work involved we recommend doing it in phases, targeting core plugins first, as defined in Essentials.
+
+===== BoM Format
+
+A new yaml format based on https://github.com/jenkins-infra/evergreen/blob/master/essentials.yaml[essentials] and https://github.com/oleg-nenashev/jenkins-war-packager[WAR packager] using the Kubernetes format.
+
+```yaml
+metadata:
+  # labels and annotations are key: value string attributes
+  labels:
+    name: myplugin
+    groupId: ...
+    artifactId: ...
+    version: ...
+  annotations:
+    io.jenkins.x: y
+spec:
+  core:
+    # version OR version + ref (version just to keep Maven happy about version order)
+    ref: master
+    version: 1.0
+  components:
+    - groupId: org.acme
+      artifactId: acme-component
+      ref: master
+      version: 1.0
+  plugins:
+    - groupId: org.acme
+      artifactId: acme-plugin
+      ref: master
+      # version: 1.0
```

Comment:
After some work on [JENKINS-50953](https://issues.jenkins-ci.org/browse/JENKINS-50953), I have a suggestion here.

First, ditch the concept of “realized BOM”. This is an aspect of my broader comment made earlier about deterministic builds. So for sake of argument, let us say there is just one BOM format, and it pins particular versions of every component.

Let us also say that the version is simply a Maven version string, as opposed to the URLs suggested by @rtyler which seems overdetermined:
* makes it harder to apply caches and proxies
* for any given product build we already have a well-known list of repositories to look in
* the update tool needs to know that same list of repositories anyway
* more verbose and error-prone

Then you could have several variants. For a plain old trunk release from `maven-release-plugin` (“MRP”):

```yaml
- groupId: io.jenkins.plugins
  artifactId: stuff
  version: 1.23
```

There is an implied `branch: master` attribute here which is elided for brevity. For an incremental release from trunk:

```yaml
- groupId: io.jenkins.plugins
  artifactId: stuff
  version: 1.24-rc9999.abc123def456
```

For an incremental release from an _unmerged_ PR:

```yaml
- groupId: io.jenkins.plugins
  artifactId: stuff
  version: 1.24-rc9999.abc123def456
  branch: jglick:refactoring-JENKINS-56789
```

(Note the syntax matching [GitHub conventions](https://developer.github.com/v3/repos/commits/#compare-two-commits).) Even other options are possible, like MRP from a backport branch (in the origin repo):

```yaml
- groupId: io.jenkins.plugins
  artifactId: stuff
  version: 1.23.1
  branch: stable
```

Wait, you exclaim! If we are pinning a `version`, why do we need a `branch` too? Two reasons.

First, the version update tool needs to know which branch to inspect for candidate versions. Remember that while incremental version numbers are meaningfully comparable within a branch, that is not true between branches, since commit history can be a DAG, so it cannot simply pick the deployed release with the highest `$(git rev-list --first-parent --count HEAD)`; it has to filter them by branch. If we want to be able to run the tool in a mode where it offers updates to every component in the file without further ado, it needs to know if there are some which are currently being taken from something other than `master`.

Related to that, we ought to be performing automated validation if this BOM is being used for production, such as the Essentials manifest: a CI build of the BOM repo could use a subset of the code in JENKINS-50953 to check that a requested incremental `version` indeed maps to something already pushed to the designated branch. As alluded to in JEP-305, this is important for security. Suppose you receive a pull request that says

```diff
-  version: 2.13-rc735.f7ae668c009b
+  version: 2.13-rc737.789e66b233dd
```

Sure, that looks OK, right? A plugin update, go for it. But how do you know what this `789e66b233dd` commit actually was? Something you trust? Are you going to construct the GitHub `commit` URL for it and take a look? Probably not, you will just merge and forget. But that might have been malware built from some fly-by-night PR which was closed the moment its artifacts got deployed to Incrementals! By contrast, with the `branch` system that PR would be marked as a build failure, since the commit is not an ancestor of the current `master`, and if the pull request looks like

```diff
-  version: 2.13-rc735.f7ae668c009b
+  version: 2.13-rc737.789e66b233dd
+  branch: l33thaxor:muahaha
```

you are going to think twice before merging it.

(A particular BOM like the Essentials manifest may of course wish to use only `master` commits as a matter of policy, and it might also be receiving routine updates from a bot rather than PRs filed by actual people, but it would be best if there is a single format that can be used safely in different contexts.)